lampshade.go

```diff
@@ -295,13 +295,16 @@ type StatsTracking interface {
 	EMARTT() time.Duration
 }
 
+// DialFN is a function that dials the server
+type DialFN func() (net.Conn, error)
+
 // Dialer is a dialer
 type Dialer interface {
 	StatsTracking
 
-	Dial() (net.Conn, error)
+	Dial(dial DialFN) (net.Conn, error)
```

Comment:
It still feels a bit uncomfortable to me, as the `dial` is expected to be underlie dialer being called each time, but it is actually `dialNewSession` and called seldomly.

Is dialing new session really part of the original op? Shouldn't it be transparent enough?

Maybe I lacks some background, but I think it's better as part of `DialerOpts` with separate op. The context can still be attached to the error in case if it fails to dial physical connection.